pkg/itemblock/actions/pvc_action.go

```diff
@@ -0,0 +1,112 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actions
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1api "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/velero/pkg/actionhelpers"
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/client"
+	"github.com/vmware-tanzu/velero/pkg/kuberesource"
+	plugincommon "github.com/vmware-tanzu/velero/pkg/plugin/framework/common"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"github.com/vmware-tanzu/velero/pkg/util/kube"
+)
+
+// PVCAction inspects a PersistentVolumeClaim for the PersistentVolume
+// that it references and backs it up
+type PVCAction struct {
+	log      logrus.FieldLogger
+	crClient crclient.Client
+}
+
+func NewPVCAction(f client.Factory) plugincommon.HandlerInitializer {
+	return func(logger logrus.FieldLogger) (interface{}, error) {
+		crClient, err := f.KubebuilderClient()
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		return &PVCAction{
+			log:      logger,
+			crClient: crClient,
+		}, nil
+	}
+}
+
+func (a *PVCAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"persistentvolumeclaims"},
+	}, nil
+}
+
+func (a *PVCAction) GetRelatedItems(item runtime.Unstructured, backup *v1.Backup) ([]velero.ResourceIdentifier, error) {
+	a.log.Info("Executing PVC ItemBlockAction")
+	defer a.log.Info("Done executing PVC ItemBlockAction")
+
+	pvc := new(corev1api.PersistentVolumeClaim)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &pvc); err != nil {
+		return nil, errors.Wrap(err, "unable to convert unstructured item to persistent volume claim")
+	}
+
+	if pvc.Status.Phase != corev1api.ClaimBound || pvc.Spec.VolumeName == "" {
+		return nil, nil
+	}
+	// returns the PV for the PVC (shared with BIA additionalItems)
+	relatedItems := actionhelpers.RelatedItemsForPVC(pvc, a.log)
+
+	// Adds pods mounting this PVC to ensure that multiple pods mounting the same RWX
+	// volume get backed up together.
+	pods := new(corev1api.PodList)
+	err := a.crClient.List(context.Background(), pods, crclient.InNamespace(pvc.Namespace))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list pods")
+	}
+
+	for i := range pods.Items {
+		if kube.IsPodRunning(&pods.Items[i]) != nil {
```

Comment:
Since the pod is dynamically enumerating from the api server, what is the consequence once below cases happen?

- After calling this  `GetRelatedItems`, a new pod starts and consumes this PVC
- By the time of calling `GetRelatedItems`, a pod consuming this PVC is filtered out for any reason (e.g., pod is not running), but later the pod just turns to running state
- ...

For all the above cases, one more more pods will not be reported by `GetRelatedItems`, then will there be some logical error in the execution of the following item block processing?

Reply:
If there is no logical error, I think the execution will be different if one or more pods are filtered here right? If so, we may need to add some logs to help the troubleshooting

Reply:
If a new pod starts after the item collector runs, that pod will probably not be in the list returned by the item collector, so it will likely not be backed up in any case.

In the not-running-to-running state, since we've already grabbed the item from the cluster during item collection, velero won't consider that new state when backing up anyway -- without a running pod, we can't run hooks anyway (which is the main reason to add these pods to the ItemBlock).

More generally, there is always a delay with velero between item collection and item backup -- if things are created or destroyed (or even modified) in-between these times, Velero doesn't make any guarantees as to what happens there. If an item is deleted after item collection, it will probably be backed up since we have the content already, although it may fail if there are BIAs that try to access the in-cluster resource. If an item is created after item collection, then it most likely won't be backed up anyway. If an item is modified after item collection, those changes generally won't make it into the backup later. The only exceptions to the above would be for items that weren't in the original item collection but added as an additional item from a BIA plugin (or, with this new functionality, an IBA plugin). In this case, we have to grab the item at the time it's returned, so if the change happened between item collection and now, we'd get the later state.

Reply:
> The only exceptions to the above would be for items that weren't in the original item collection but added as an additional item from a BIA plugin (or, with this new functionality, an IBA plugin). In this case, we have to grab the item at the time it's returned, so if the change happened between item collection and now, we'd get the later state.

In this case, backing up the old state or the new state are both expected and fine, but is there any other impact to the logic of BIA and IBA since the object is missing in IBA, e.g., will this case the item being backed up twice?

Reply:
OK. And do we need a log here since this will affect the behavior of itemblock backup?
E.g., for a RWX PVC, if the pod is not running, `GetRelatedItems` won't return the pod; then the pod turns to running state, so Velero backs it up; so there may be two pods being backed up. Since the two pods may not be in the same itemblock in this case, will the PVC be backed up twice?

Reply:
"then the pod turns to running state, so Velero backs it up"
State changes between item collection and item backup won't be used by velero, since Velero obtains all pod data via per-namespace `List` calls obtained by the item collector. It's probably a good idea to add a log message "not returning non-running pod" or something, though. I can add that.

If non-running pod scenarios did end up causing the second pod to *not* be in the same ItemBlock, the PVC won't be backed up twice because the ItemBlock creation logic always checks returned related items -- if it's already in an ItemBlock, then it won't be added to a second one.